File: static/python/airtable_runtime.py
"""
AirtableRuntime: Implements Airtable formula functions and operators in Python.

All operators route through this class for correct Airtable semantics:
- BLANK() is None (treated as 0 in numeric context, "" in string context)
- Division by zero returns float('nan')
- Type coercion follows Airtable rules
"""

from __future__ import annotations

import math
import re
from datetime import datetime, timedelta
from typing import Any


class AirtableRuntime:
    # region Utilities
    @staticmethod
    def _is_blank(v: Any) -> bool:
        return v is None

    @staticmethod
    def N(v: Any) -> int | float:  # noqa: N802
        """Coerce value to number"""
        if isinstance(v, list):
            return AirtableRuntime.N(v[0] if v else None)
        if v is None:
            return 0
        if isinstance(v, bool):
            return 1 if v else 0
        if isinstance(v, (int, float)):
            return v
        if isinstance(v, str):
            try:
                return float(v) if "." in v else int(v)
            except (ValueError, TypeError):
                return 0
        return 0

    @staticmethod
    def S(v: Any) -> str:  # noqa: N802
        """Coerce value to string"""
        if isinstance(v, list):
            return AirtableRuntime.S(v[0] if v else None)
        if v is None:
            return ""
        if isinstance(v, bool):
            return "1" if v else "0"
        return str(v)

    @staticmethod
    def _flat_args(args: tuple[Any, ...]) -> list[Any]:
        result: list[Any] = []
        for a in args:
            if isinstance(a, list):
                result.extend(a)
            else:
                result.append(a)
        return result

    # endregion

    # region Numeric functions
    @staticmethod
    def SUM(*args: Any) -> int | float:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        return sum(AirtableRuntime.N(v) for v in flat)

    @staticmethod
    def AVERAGE(*args: Any) -> float:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        if not flat:
            return float("nan")
        return AirtableRuntime.SUM(*flat) / len(flat)

    @staticmethod
    def MIN(*args: Any) -> int | float:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        if not flat:
            return float("inf")
        return min(AirtableRuntime.N(v) for v in flat)

    @staticmethod
    def MAX(*args: Any) -> int | float:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        if not flat:
            return float("-inf")
        return max(AirtableRuntime.N(v) for v in flat)

    @staticmethod
    def COUNT(*args: Any) -> int:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        return sum(1 for v in flat if isinstance(v, (int, float)) and not (isinstance(v, float) and math.isnan(v)))

    @staticmethod
    def COUNTA(*args: Any) -> int:  # noqa: N802
        flat = AirtableRuntime._flat_args(args)
        return sum(1 for v in flat if not AirtableRuntime._is_blank(v) and v != "")

    @staticmethod
    def COUNTALL(*args: Any) -> int:  # noqa: N802
        return len(AirtableRuntime._flat_args(args))

    @staticmethod
    def ROUND(value: Any, precision: Any = 0) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        p = int(AirtableRuntime.N(precision))
        return round(n, p)

    @staticmethod
    def ROUNDUP(value: Any, precision: Any = 0) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        p = int(AirtableRuntime.N(precision))
        factor = 10**p
        return math.ceil(n * factor) / factor

    @staticmethod
    def ROUNDDOWN(value: Any, precision: Any = 0) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        p = int(AirtableRuntime.N(precision))
        factor = 10**p
        return math.floor(n * factor) / factor

    @staticmethod
    def CEILING(value: Any, significance: Any = 1) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        s = AirtableRuntime.N(significance) or 1
        return math.ceil(n / s) * s

    @staticmethod
    def FLOOR(value: Any, significance: Any = 1) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        s = AirtableRuntime.N(significance) or 1
        return math.floor(n / s) * s

    @staticmethod
    def POWER(base: Any, exponent: Any) -> float:  # noqa: N802
        return math.pow(AirtableRuntime.N(base), AirtableRuntime.N(exponent))

    @staticmethod
    def EXP(value: Any) -> float:  # noqa: N802
        return math.exp(AirtableRuntime.N(value))

    @staticmethod
    def LOG(value: Any, base: Any = None) -> float:  # noqa: N802
        n = AirtableRuntime.N(value)
        if base is None:
            return math.log10(n)
        return math.log(n) / math.log(AirtableRuntime.N(base))

    @staticmethod
    def LOG10(value: Any) -> float:  # noqa: N802
        return math.log10(AirtableRuntime.N(value))

    @staticmethod
    def MOD(value: Any, divisor: Any) -> int | float:  # noqa: N802
        n = AirtableRuntime.N(value)
        d = AirtableRuntime.N(divisor)
        if d == 0:
            return float("nan")
        return n % d

    @staticmethod
    def EVEN(value: Any) -> int:  # noqa: N802
        n = AirtableRuntime.N(value)
        ceil_val = math.ceil(abs(n))
        result = ceil_val if ceil_val % 2 == 0 else ceil_val + 1
        return -result if n < 0 else result

    @staticmethod
    def ODD(value: Any) -> int:  # noqa: N802
        n = AirtableRuntime.N(value)
        ceil_val = math.ceil(abs(n))
        result = ceil_val if ceil_val % 2 == 1 else ceil_val + 1
        return -result if n < 0 else result

    @staticmethod
    def VALUE(value: Any) -> int | float:  # noqa: N802
        if AirtableRuntime._is_blank(value):
            return 0
        try:
            s = str(value)
            return float(s) if "." in s else int(s)
        except (ValueError, TypeError):
            return float("nan")

    # endregion

    # region String functions
    @staticmethod
    def CONCATENATE(*args: Any) -> str:  # noqa: N802
        return "".join(AirtableRuntime.S(a) for a in args)

    @staticmethod
    def LEFT(text: Any, count: Any) -> str:  # noqa: N802
        return AirtableRuntime.S(text)[: int(AirtableRuntime.N(count))]

    @staticmethod
    def RIGHT(text: Any, count: Any) -> str:  # noqa: N802
        s = AirtableRuntime.S(text)
        n = int(AirtableRuntime.N(count))
        return s[max(0, len(s) - n) :]

    @staticmethod
    def MID(text: Any, start: Any, count: Any) -> str:  # noqa: N802
        s = AirtableRuntime.S(text)
        start_idx = int(AirtableRuntime.N(start)) - 1
        length = int(AirtableRuntime.N(count))
        return s[start_idx : start_idx + length]

    @staticmethod
    def FIND(needle: Any, haystack: Any, start: Any = None) -> int:  # noqa: N802
        s = AirtableRuntime.S(haystack)
        n = AirtableRuntime.S(needle)
        start_idx = 0 if AirtableRuntime._is_blank(start) else int(AirtableRuntime.N(start)) - 1
        idx = s.find(n, start_idx)
        return 0 if idx == -1 else idx + 1

    @staticmethod
    def SEARCH(needle: Any, haystack: Any, start: Any = None) -> int:  # noqa: N802
        s = AirtableRuntime.S(haystack).lower()
        n = AirtableRuntime.S(needle).lower()
        start_idx = 0 if AirtableRuntime._is_blank(start) else int(AirtableRuntime.N(start)) - 1
        idx = s.find(n, start_idx)
        return 0 if idx == -1 else idx + 1

    @staticmethod
    def SUBSTITUTE(text: Any, old_str: Any, new_str: Any, index: Any = None) -> str:  # noqa: N802
        s = AirtableRuntime.S(text)
        o = AirtableRuntime.S(old_str)
        n = AirtableRuntime.S(new_str)
        if AirtableRuntime._is_blank(index):
            return s.replace(o, n)
        target = int(AirtableRuntime.N(index))
        count = 0

        def _replacer(match: re.Match[str]) -> str:
            nonlocal count
            count += 1
            return n if count == target else match.group()

        return re.sub(re.escape(o), _replacer, s)

    @staticmethod
    def REPLACE(text: Any, start: Any, count: Any, replacement: Any) -> str:  # noqa: N802
        s = AirtableRuntime.S(text)
        start_idx = int(AirtableRuntime.N(start)) - 1
        length = int(AirtableRuntime.N(count))
        return s[:start_idx] + AirtableRuntime.S(replacement) + s[start_idx + length :]

    @staticmethod
    def REPT(text: Any, count: Any) -> str:  # noqa: N802
        return AirtableRuntime.S(text) * max(0, int(AirtableRuntime.N(count)))

    @staticmethod
    def T(value: Any) -> str:  # noqa: N802
        return value if isinstance(value, str) else ""

    @staticmethod
    def REGEX_MATCH(text: Any, regex: Any) -> bool:  # noqa: N802
        try:
            return bool(re.search(AirtableRuntime.S(regex), AirtableRuntime.S(text)))
        except re.error:
            return False

    @staticmethod
    def REGEX_EXTRACT(text: Any, regex: Any) -> str | None:  # noqa: N802
        try:
            match = re.search(AirtableRuntime.S(regex), AirtableRuntime.S(text))
            return match.group(0) if match else None
        except re.error:
            return None

    @staticmethod
    def REGEX_REPLACE(text: Any, regex: Any, replacement: Any) -> str:  # noqa: N802
        try:
            return re.sub(AirtableRuntime.S(regex), AirtableRuntime.S(replacement), AirtableRuntime.S(text))
        except re.error:
            return AirtableRuntime.S(text)

    # endregion

    # region Date/Time functions
    @staticmethod
    def TODAY() -> str:  # noqa: N802
        return datetime.now().strftime("%Y-%m-%d")

    @staticmethod
    def NOW() -> str:  # noqa: N802
        return datetime.now().isoformat()

    @staticmethod
    def DATEADD(date: Any, count: Any, unit: Any) -> str | None:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return None
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return None
        n = int(AirtableRuntime.N(count))
        u = AirtableRuntime.S(unit).lower()
        if u == "years":
            d = d.replace(year=d.year + n)
        elif u == "months":
            month = d.month + n
            year = d.year + (month - 1) // 12
            month = (month - 1) % 12 + 1
            d = d.replace(year=year, month=month)
        elif u == "weeks":
            d += timedelta(weeks=n)
        elif u == "days":
            d += timedelta(days=n)
        elif u == "hours":
            d += timedelta(hours=n)
        elif u == "minutes":
            d += timedelta(minutes=n)
        elif u == "seconds":
            d += timedelta(seconds=n)
        return d.isoformat()

    @staticmethod
    def DATETIME_DIFF(date1: Any, date2: Any, unit: Any = None) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date1) or AirtableRuntime._is_blank(date2):
            return 0
        try:
            d1 = datetime.fromisoformat(AirtableRuntime.S(date1).replace("Z", "+00:00"))
            d2 = datetime.fromisoformat(AirtableRuntime.S(date2).replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return 0
        u = AirtableRuntime.S(unit or "days").lower()
        diff = d1 - d2
        total_seconds = diff.total_seconds()
        if u == "milliseconds":
            return int(total_seconds * 1000)
        elif u == "seconds":
            return int(total_seconds)
        elif u == "minutes":
            return int(total_seconds / 60)
        elif u == "hours":
            return int(total_seconds / 3600)
        elif u == "days":
            return int(total_seconds / 86400)
        elif u == "weeks":
            return int(total_seconds / (86400 * 7))
        elif u == "months":
            return (d1.year - d2.year) * 12 + (d1.month - d2.month)
        elif u == "years":
            return d1.year - d2.year
        return int(total_seconds / 86400)

    @staticmethod
    def DATETIME_FORMAT(date: Any, _format: Any = None) -> str:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return ""
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
            return d.isoformat()
        except (ValueError, TypeError):
            return ""

    @staticmethod
    def DATETIME_PARSE(text: Any, _format: Any = None, _locale: Any = None) -> str | None:  # noqa: N802
        if AirtableRuntime._is_blank(text):
            return None
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(text).replace("Z", "+00:00"))
            return d.isoformat()
        except (ValueError, TypeError):
            return None

    @staticmethod
    def SET_LOCALE(date: Any, _locale: Any) -> Any:  # noqa: N802
        return date

    @staticmethod
    def SET_TIMEZONE(date: Any, _timezone: Any) -> Any:  # noqa: N802
        return date

    @staticmethod
    def YEAR(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).year
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def MONTH(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).month
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def DAY(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).day
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def HOUR(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).hour
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def MINUTE(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).minute
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def SECOND(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            return datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00")).second
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def WEEKDAY(date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
            return (d.weekday() + 1) % 7  # Sunday=0, Monday=1, ...
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def WEEKNUM(date: Any, start_day: Any = None) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return 0
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
            day_names = {"sunday": 6, "monday": 0, "tuesday": 1, "wednesday": 2, "thursday": 3, "friday": 4, "saturday": 5}
            if AirtableRuntime._is_blank(start_day):
                start_dow = 6  # Sunday (Python: Monday=0)
            else:
                start_dow = day_names.get(AirtableRuntime.S(start_day).lower(), 6)
            start_of_year = datetime(d.year, 1, 1, tzinfo=d.tzinfo)
            day_of_year = (d - start_of_year).days
            start_day_of_week = start_of_year.weekday()  # Monday=0
            adjusted = day_of_year + ((start_day_of_week - start_dow + 7) % 7)
            return (adjusted // 7) + 1
        except (ValueError, TypeError):
            return 0

    @staticmethod
    def DATESTR(date: Any) -> str:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return ""
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
            return d.strftime("%Y-%m-%d")
        except (ValueError, TypeError):
            return ""

    @staticmethod
    def TIMESTR(date: Any) -> str:  # noqa: N802
        if AirtableRuntime._is_blank(date):
            return ""
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(date).replace("Z", "+00:00"))
            return d.strftime("%H:%M:%S")
        except (ValueError, TypeError):
            return ""

    @staticmethod
    def TONOW(date: Any, unit: Any = None) -> int:  # noqa: N802
        return AirtableRuntime.DATETIME_DIFF(datetime.now().isoformat(), date, unit or "days")

    @staticmethod
    def FROMNOW(date: Any, unit: Any = None) -> int:  # noqa: N802
        return AirtableRuntime.DATETIME_DIFF(date, datetime.now().isoformat(), unit or "days")

    @staticmethod
    def IS_SAME(date1: Any, date2: Any, unit: Any = None) -> bool:  # noqa: N802
        return AirtableRuntime.DATETIME_DIFF(date1, date2, unit or "days") == 0

    @staticmethod
    def IS_BEFORE(date1: Any, date2: Any, unit: Any = None) -> bool:  # noqa: N802
        return AirtableRuntime.DATETIME_DIFF(date1, date2, unit or "days") < 0

    @staticmethod
    def IS_AFTER(date1: Any, date2: Any, unit: Any = None) -> bool:  # noqa: N802
        return AirtableRuntime.DATETIME_DIFF(date1, date2, unit or "days") > 0

    @staticmethod
    def WORKDAY(start_date: Any, num_days: Any) -> str | None:  # noqa: N802
        if AirtableRuntime._is_blank(start_date):
            return None
        try:
            d = datetime.fromisoformat(AirtableRuntime.S(start_date).replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return None
        remaining = int(AirtableRuntime.N(num_days))
        direction = 1 if remaining > 0 else -1
        remaining = abs(remaining)
        while remaining > 0:
            d += timedelta(days=direction)
            if d.weekday() < 5:  # Monday=0 through Friday=4
                remaining -= 1
        return d.isoformat()

    @staticmethod
    def WORKDAY_DIFF(start_date: Any, end_date: Any) -> int:  # noqa: N802
        if AirtableRuntime._is_blank(start_date) or AirtableRuntime._is_blank(end_date):
            return 0
        try:
            d1 = datetime.fromisoformat(AirtableRuntime.S(start_date).replace("Z", "+00:00"))
            d2 = datetime.fromisoformat(AirtableRuntime.S(end_date).replace("Z", "+00:00"))
        except (ValueError, TypeError):
            return 0
        count = 0
        current = d1
        direction = 1 if d2 > d1 else -1
        while (direction == 1 and current < d2) or (direction == -1 and current > d2):
            current += timedelta(days=direction)
            if current.weekday() < 5:
                count += direction
        return count

    # endregion

    # region Array functions
    @staticmethod
    def ARRAYJOIN(arr: Any, separator: Any = None) -> str:  # noqa: N802
        if not isinstance(arr, list):
            return AirtableRuntime.S(arr)
        sep = ", " if AirtableRuntime._is_blank(separator) else AirtableRuntime.S(separator)
        return sep.join(AirtableRuntime.S(v) for v in arr)

    @staticmethod
    def ARRAYUNIQUE(arr: Any) -> list[Any]:  # noqa: N802
        if not isinstance(arr, list):
            return [arr]
        seen: list[Any] = []
        for v in arr:
            if v not in seen:
                seen.append(v)
        return seen

    @staticmethod
    def ARRAYCOMPACT(arr: Any) -> list[Any]:  # noqa: N802
        if not isinstance(arr, list):
            return [] if AirtableRuntime._is_blank(arr) else [arr]
        return [v for v in arr if not AirtableRuntime._is_blank(v) and v != ""]

    @staticmethod
    def ARRAYFLATTEN(arr: Any) -> list[Any]:  # noqa: N802
        if not isinstance(arr, list):
            return [arr]
        result: list[Any] = []

        def _flatten(items: list[Any]) -> None:
            for item in items:
                if isinstance(item, list):
                    _flatten(item)
                else:
                    result.append(item)

        _flatten(arr)
        return result

    # endregion

    # region Record/Special
    @staticmethod
    def ERROR(message: Any = None) -> None:  # noqa: N802
        raise ValueError(AirtableRuntime.S(message or "Error"))

    @staticmethod
    def ISERROR(value: Any) -> bool:  # noqa: N802
        return isinstance(value, Exception) or (isinstance(value, float) and math.isnan(value))

    # endregion

File: static/rust/airtable_runtime.rs
use serde_json::Value;

// =============================================================================
// Type Coercion
// =============================================================================

/// Coerce a value to f64. null→0, bool→0/1, string→parse or 0, array→first element.
#[allow(non_snake_case)]
pub fn N(v: &Value) -> f64 {
    match v {
        Value::Null => 0.0,
        Value::Bool(b) => {
            if *b {
                1.0
            } else {
                0.0
            }
        }
        Value::Number(n) => n.as_f64().unwrap_or(0.0),
        Value::String(s) => s.parse::<f64>().unwrap_or(0.0),
        Value::Array(arr) => {
            if let Some(first) = arr.first() {
                N(first)
            } else {
                0.0
            }
        }
        _ => 0.0,
    }
}

/// Coerce a value to String. null→"", bool→"0"/"1", float strips .0 for whole numbers.
#[allow(non_snake_case)]
pub fn S(v: &Value) -> String {
    match v {
        Value::Null => String::new(),
        Value::Bool(b) => {
            if *b {
                "1".to_string()
            } else {
                "0".to_string()
            }
        }
        Value::Number(n) => {
            let f = n.as_f64().unwrap_or(0.0);
            if f.is_nan() {
                "NaN".to_string()
            } else if f.is_infinite() {
                if f > 0.0 {
                    "Infinity".to_string()
                } else {
                    "-Infinity".to_string()
                }
            } else if f == f.trunc() && f.is_finite() {
                format!("{}", f as i64)
            } else {
                format!("{}", f)
            }
        }
        Value::String(s) => s.clone(),
        Value::Array(arr) => {
            if let Some(first) = arr.first() {
                S(first)
            } else {
                String::new()
            }
        }
        _ => String::new(),
    }
}

/// Flatten one level of arrays.
#[allow(non_snake_case)]
pub fn A(args: &[Value]) -> Vec<Value> {
    let mut result = Vec::new();
    for v in args {
        if let Value::Array(arr) = v {
            result.extend(arr.iter().cloned());
        } else {
            result.push(v.clone());
        }
    }
    result
}

/// Flatten one level, then coerce each to f64.
#[allow(non_snake_case)]
pub fn AN(args: &[Value]) -> Vec<f64> {
    A(args).iter().map(|v| N(v)).collect()
}

fn to_value(f: f64) -> Value {
    if f.is_nan() || f.is_infinite() {
        Value::Null
    } else if f == f.trunc() && f.abs() < (i64::MAX as f64) {
        serde_json::json!(f as i64)
    } else {
        serde_json::Number::from_f64(f)
            .map(Value::Number)
            .unwrap_or(Value::Null)
    }
}

// =============================================================================
// Math Functions
// =============================================================================

/// Sum of all values.
#[allow(non_snake_case)]
pub fn SUM(args: &[Value]) -> Value {
    to_value(AN(args).iter().sum())
}

/// Average of all values.
#[allow(non_snake_case)]
pub fn AVERAGE(args: &[Value]) -> Value {
    let nums = AN(args);
    if nums.is_empty() {
        return Value::Null; // NaN
    }
    to_value(nums.iter().sum::<f64>() / nums.len() as f64)
}

/// Count of numeric, non-NaN values.
#[allow(non_snake_case)]
pub fn COUNT(args: &[Value]) -> Value {
    let flat = A(args);
    let count = flat
        .iter()
        .filter(|v| matches!(v, Value::Number(_)) && !N(v).is_nan())
        .count();
    serde_json::json!(count)
}

/// Count of non-null, non-empty-string values.
#[allow(non_snake_case)]
pub fn COUNTA(args: &[Value]) -> Value {
    let flat = A(args);
    let count = flat
        .iter()
        .filter(|v| !v.is_null() && *v != &Value::String(String::new()))
        .count();
    serde_json::json!(count)
}

/// Minimum value.
#[allow(non_snake_case)]
pub fn MIN(args: &[Value]) -> Value {
    let nums = AN(args);
    to_value(nums.iter().copied().fold(f64::INFINITY, f64::min))
}

/// Maximum value.
#[allow(non_snake_case)]
pub fn MAX(args: &[Value]) -> Value {
    let nums = AN(args);
    to_value(nums.iter().copied().fold(f64::NEG_INFINITY, f64::max))
}

/// Round to precision decimal places.
#[allow(non_snake_case)]
pub fn ROUND(v: &Value, precision: &Value) -> Value {
    let n = N(v);
    let p = N(precision) as i32;
    let factor = 10f64.powi(p);
    to_value((n * factor).round() / factor)
}

/// Round up (away from zero) to precision decimal places.
#[allow(non_snake_case)]
pub fn ROUNDUP(v: &Value, precision: &Value) -> Value {
    let n = N(v);
    let p = N(precision) as i32;
    let factor = 10f64.powi(p);
    to_value((n * factor).ceil() / factor)
}

/// Round down (toward zero) to precision decimal places.
#[allow(non_snake_case)]
pub fn ROUNDDOWN(v: &Value, precision: &Value) -> Value {
    let n = N(v);
    let p = N(precision) as i32;
    let factor = 10f64.powi(p);
    to_value((n * factor).floor() / factor)
}

/// Round up to nearest multiple of significance.
#[allow(non_snake_case)]
pub fn CEILING(v: &Value, significance: &Value) -> Value {
    let n = N(v);
    let s = {
        let sig = N(significance);
        if sig == 0.0 {
            1.0
        } else {
            sig
        }
    };
    to_value((n / s).ceil() * s)
}

/// Round down to nearest multiple of significance.
#[allow(non_snake_case)]
pub fn FLOOR(v: &Value, significance: &Value) -> Value {
    let n = N(v);
    let s = {
        let sig = N(significance);
        if sig == 0.0 {
            1.0
        } else {
            sig
        }
    };
    to_value((n / s).floor() * s)
}

/// Logarithm. Default base 10.
#[allow(non_snake_case)]
pub fn LOG(v: &Value, base: Option<&Value>) -> Value {
    let n = N(v);
    let b = base.map(|b| N(b)).unwrap_or(10.0);
    to_value(n.ln() / b.ln())
}

/// Round up to nearest even integer.
#[allow(non_snake_case)]
pub fn EVEN(v: &Value) -> Value {
    let n = N(v);
    let sign = if n < 0.0 { -1.0 } else { 1.0 };
    let abs_ceil = n.abs().ceil() as i64;
    let result = if abs_ceil % 2 == 0 {
        abs_ceil
    } else {
        abs_ceil + 1
    };
    to_value(sign * result as f64)
}

/// Round up to nearest odd integer.
#[allow(non_snake_case)]
pub fn ODD(v: &Value) -> Value {
    let n = N(v);
    let sign = if n < 0.0 { -1.0 } else { 1.0 };
    let abs_ceil = n.abs().ceil() as i64;
    let result = if abs_ceil % 2 == 1 {
        abs_ceil
    } else {
        abs_ceil + 1
    };
    to_value(sign * result as f64)
}

/// Parse string to number. Invalid → null (NaN).
#[allow(non_snake_case)]
pub fn VALUE(v: &Value) -> Value {
    match v {
        Value::Null => to_value(0.0),
        Value::Number(_) => v.clone(),
        Value::String(s) => {
            match s.parse::<f64>() {
                Ok(n) => to_value(n),
                Err(_) => Value::Null, // NaN
            }
        }
        _ => Value::Null,
    }
}

/// Power.
#[allow(non_snake_case)]
pub fn POWER(base: &Value, exp: &Value) -> Value {
    to_value(N(base).powf(N(exp)))
}

/// Modulo.
#[allow(non_snake_case)]
pub fn MOD(v: &Value, divisor: &Value) -> Value {
    to_value(N(v) % N(divisor))
}

/// Absolute value.
#[allow(non_snake_case)]
pub fn ABS(v: &Value) -> Value {
    to_value(N(v).abs())
}

/// Square root.
#[allow(non_snake_case)]
pub fn SQRT(v: &Value) -> Value {
    to_value(N(v).sqrt())
}

/// Truncate to integer.
#[allow(non_snake_case)]
pub fn INT(v: &Value) -> Value {
    to_value(N(v).trunc())
}

// =============================================================================
// Logic Functions
// =============================================================================

/// Conditional: if condition is truthy, return if_true, else if_false.
#[allow(non_snake_case)]
pub fn IF(condition: &Value, if_true: &Value, if_false: &Value) -> Value {
    if is_truthy(condition) {
        if_true.clone()
    } else {
        if_false.clone()
    }
}

/// Switch/case expression.
#[allow(non_snake_case)]
pub fn SWITCH(expr: &Value, cases: &[(Value, Value)], default: Option<&Value>) -> Value {
    for (pattern, result) in cases {
        if expr == pattern {
            return result.clone();
        }
    }
    default.cloned().unwrap_or(Value::Null)
}

/// Return null (BLANK).
#[allow(non_snake_case)]
pub fn BLANK() -> Value {
    Value::Null
}

/// Return an error marker (null, since JSON can't represent errors).
#[allow(non_snake_case)]
pub fn ERROR(_message: Option<&Value>) -> Value {
    Value::Null
}

/// Check if value is an error (NaN or null from error context).
#[allow(non_snake_case)]
pub fn ISERROR(v: &Value) -> Value {
    Value::Bool(v.is_null())
}

/// Return true.
#[allow(non_snake_case)]
pub fn TRUE() -> Value {
    Value::Bool(true)
}

/// Return false.
#[allow(non_snake_case)]
pub fn FALSE() -> Value {
    Value::Bool(false)
}

/// Check if a value is truthy (non-null, non-false, non-zero, non-empty-string).
pub fn is_truthy(v: &Value) -> bool {
    match v {
        Value::Null => false,
        Value::Bool(b) => *b,
        Value::Number(n) => n.as_f64().unwrap_or(0.0) != 0.0,
        Value::String(s) => !s.is_empty(),
        Value::Array(arr) => !arr.is_empty(),
        Value::Object(_) => true,
    }
}
